library/ssl_tls12_server.c

```diff
@@ -3946,7 +3946,7 @@ static int ssl_parse_client_key_exchange(mbedtls_ssl_context *ssl)
            The size of the array is thus greater than 256 bytes which is greater than any
            possible value of ecpoint_len (type uint8_t) and the check below can be skipped.*/
```

Comment:
When FFDH is disabled but RSA is enabled, `sizeof(handshake->xxdh_psa_peerkey)` is also greater than 256 bytes (assuming a sensible maximum RSA key size — if I got the math right, setting `PSA_VENDOR_RSA_MAX_KEY_BITS` to 1912 or less would result in ≤255 bytes for `sizeof(handshake->xxdh_psa_peerkey)` which is `PSA_EXPORT_PUBLIC_KEY_MAX_SIZE`).

This doesn't need to be the case, since RSA is irrelevant to what's stored in `xxdh_psa_peerkey`, but the size of the array is declared as `PSA_EXPORT_PUBLIC_KEY_MAX_SIZE` and not as max(max FFDH public key size, max ECC public key size).

Skipping this check when FFDH is disabled was done in 24e50d3dbdd1893364c241c179dbf182f2bf9413 to silence this compiler warning. We could broaden the skip to whenever the buffer is larger than `UINT8_MAX`. But I don't like the way this is done: I'd prefer to let compilers figure out if the comparison is always true and optimize the branch out. Evidently Clang, at least, figures it out.

I'd like a second opinion from another Mbed TLS maintainer on the preferred fix here.

Reply:
Thanks for the explanation. It's interesting to learn a bit about the internals of the library. Without knowing all the implications, I'd in general prefer as well to let compilers figure out the comparison. This already came in handy here with Clang forcing me to look at potential mistakes without letting them slip through.